server/src/engine/queue.ts

```diff
@@ -69,7 +69,36 @@ function handleEnqueue() {
 }
 
 function makeEnqueueDebounced() {
-  return _.debounce(engine.compileOnChangeDelay(), handleEnqueue, { leading: true, trailing: true })
+  let coolDownTimer: NodeJS.Timeout | null = null
+  let waiting = false
+
+  function startCooldown() {
+    coolDownTimer = setTimeout(endCooldown, engine.compileOnChangeDelay())
+  }
+
+  function endCooldown() {
+    coolDownTimer = null
+
+    // If there was a call during cool-down, handle it now
+    if (waiting) {
+      waiting = false
+      handleEnqueue()
+      startCooldown()
+    }
+  }
+
+  return () => {
+    // If we're still in cool-down,
+    // register that we want to be called at the end of the cool-down
+    if (coolDownTimer !== null) {
+      waiting = true
+      return
+    }
+
+    // If we're not in cool down, call immediately and start cool-down
```

Comment:
I don't think this is the right behavior we expect from debounce.

The point is that when you are typing we want to wait DELAY ms after _the last keystroke_. 

We need to wait for the user to stop typing and then act. 

We should NOT act on the first keystroke.

Reply:
I think the correct behavior is to clear any existing timer and then schedule a new one.

Reply:
Wouldn't that completely kill auto completions?

Reply:
Is this the way it works now?

Reply:
What does the debounce do then?

Reply:
Not much
See #318

Reply:
I have tested this implementation while it was enabled though

Reply:
Hmm. I see.

Reply:
I am not sure what to do until I know how the new parser works.

Reply:
To be clear, I understand that this implementation isn't equivalent to the `_.debounce()`. This is more of a rate limit.

I'm just questioning whether the old one would have worked if we were to enable it with the auto completions added.

Reply:
So we have compileOnChangeDelay, but you are saying this value is always zero?

Reply:
On the change event we are passing in a `skipDelay` flag that circumvents the debouncing.

Reply:
Roger.